### PR TITLE
added energon dataloader for neva training

### PR DESCRIPTION
**Pull request by yashaswikarnati**
# What does this PR do ?
added megatron-energon  based datamodule for neva training

**Collection**: [Note which collection this PR will affect]
Multimodal
# Changelog 
- Energon based 

```python
from nemo.collections.multimodal.data.energon import (
    ImageToken,
    MultiModalSampleConfig,
    SimpleMultiModalDataModule,
)
processor = AutoProcessor.from_pretrained("llava-hf/llava-1.5-7b-hf")
tokenizer = processor.tokenizer
image_processor = processor.image_processor

config = MultiModalSampleConfig(
            image_token=ImageToken(token_str="<image>", token_id=-200), ignore_place_holder=-100
        )
       
data_module = SimpleMultiModalDataModule(
            path=dataset_path,
            tokenizer=tokenizer,
            image_processor=image_processor,
            num_workers=0,
            micro_batch_size=1,
            multimodal_sample_config=config,
        )
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/multimodal/data/energon/base.py [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+        super().__init__()
+        self.path = path
+        self.tokenizer = tokenizer
+        self.image_processor = image_processor
+        self.seq_length = seq_length
+        self.micro_batch_size = micro_batch_size
+        self.num_workers = num_workers
+        self.pin_memory = pin_memory
+        self.multimodal_sample_config = multimodal_sample_config
+        self.task_encoder = MultiModalTaskEncoder(
```
Comment (ethanhe42) [inline nemo/collections/multimodal/data/energon/base.py:95]: this is specific to image understanding task. could we maybe make this more generic? like user should provide the taskencoder.

File: nemo/collections/multimodal/data/energon/base.py [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+
+        This method initializes the micro-batch calculator based on the global rank and batch size.
+        It ensures that the data is properly distributed across multiple devices in a distributed setting.
+
+        Parameters:
+        global_rank (int): The global rank of the current process in the distributed training setup.
+        """
+        from nemo.lightning.data import setup_microbatch_calculator
+
+        self.global_batch_size = self.micro_batch_size * parallel_state.get_data_parallel_world_size()
```
Comment (ethanhe42) [inline nemo/collections/multimodal/data/energon/base.py:344]: allow user to set global_batch_size? some might want to use something larger for grad accumulation

File: nemo/collections/multimodal/data/energon/base.py (hunk 1 above)
Comment (yashaswikarnati) [reply nemo/collections/multimodal/data/energon/base.py]: refactored to have the user provide a custom task encoder, if not specified default one will be used.

File: nemo/collections/multimodal/data/energon/base.py (hunk 2 above)
Comment (yashaswikarnati) [reply nemo/collections/multimodal/data/energon/base.py]: Thanks for point out. Allowed the user to specify custom global batch sizes for grad accumulation.

**Review by ethanhe42 — APPROVED** [review]

**Review by ethanhe42 — APPROVED** [review]
lgtm

File: examples/multimodal/multimodal_llm/neva/neva_pretrain_energon.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
```
Comment (yaoyu-33) [inline examples/multimodal/multimodal_llm/neva/neva_pretrain_energon.py:1]: plz hide this example for now, nemo run team only wants to display nemo-run examples, unless the script is used for tests

File: nemo/collections/multimodal/data/energon/taskencoder.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    InterleavedSample,
+    SimilarityInterleavedSample,
+    VQASample,
+    batch_list,
+    batch_pad_stack,
+)
+
+from nemo.utils import logging
+
+from .config import ImageTextRawBatch, ImageTextSample
```
Comment (yaoyu-33) [inline nemo/collections/multimodal/data/energon/taskencoder.py:31]: NeMo avoids using relative importing, can we update to absolute path?

Comment (yaoyu-33) [reply nemo/collections/multimodal/data/energon/taskencoder.py:31]: Same below

File: tests/collections/multimodal/data/energon/test_task_encoder.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (yaoyu-33) [inline tests/collections/multimodal/data/energon/test_task_encoder.py:1]: update all 2023 to 2024 in the license

File: nemo/collections/multimodal/data/energon/sampleencoder.py [hunk 6]
```diff
@@ -0,0 +1,528 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (yaoyu-33) [inline nemo/collections/multimodal/data/energon/sampleencoder.py:1]: is it better to call them `sample_encoder.py` and `task_encoder.py`

File: nemo/collections/multimodal/data/energon/sampleencoder.py [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+        ImageTextSample: The encoded sample stored in output_sample.
+        """
+        # apply prompt template
+        conversation_prompt = self.apply_prompt_template(input_sample)
+        logging.debug(f"task encoder encode_sample conversation_prompt {conversation_prompt}")
+        # tokenize prompt
+        tokens = self.tokenize(conversation_prompt)
+        labels = self.compute_labels(tokens, input_sample)
+
+        tokens = tokens[:-1].contiguous()
```
Comment (yaoyu-33) [inline nemo/collections/multimodal/data/energon/sampleencoder.py:311]: is it necessary to call contiguous? just wonder

File: nemo/collections/multimodal/data/energon/taskencoder.py (hunk 4 above)
Comment (yashaswikarnati) [reply nemo/collections/multimodal/data/energon/taskencoder.py]: done

File: examples/multimodal/multimodal_llm/neva/neva_pretrain_energon.py (hunk 3 above)
Comment (yashaswikarnati) [reply examples/multimodal/multimodal_llm/neva/neva_pretrain_energon.py]: done

File: tests/collections/multimodal/data/energon/test_task_encoder.py (hunk 5 above)
Comment (yashaswikarnati) [reply tests/collections/multimodal/data/energon/test_task_encoder.py]: done

File: nemo/collections/multimodal/data/energon/sampleencoder.py (hunk 6 above)
Comment (yashaswikarnati) [reply nemo/collections/multimodal/data/energon/sampleencoder.py]: done

File: nemo/collections/multimodal/data/energon/sampleencoder.py (hunk 7 above)
Comment (yashaswikarnati) [reply nemo/collections/multimodal/data/energon/sampleencoder.py]: not really necessary, I can just leave it as well.

File: nemo/collections/multimodal/data/energon/config.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    #: Input images (N, C, H, W)
+    images: torch.Tensor = field(default_factory=lambda: torch.empty(0))
+    #: Context string
+    tokens: torch.Tensor = field(default_factory=lambda: torch.empty(0, dtype=torch.long))
+    labels: torch.Tensor = field(default_factory=lambda: torch.empty(0, dtype=torch.long))
+    loss_mask: torch.Tensor = field(default_factory=lambda: torch.empty(0, dtype=torch.float))
+
+
+@dataclass
+class ConversationTemplateConfig:
```
Comment (yaoyu-33) [inline nemo/collections/multimodal/data/energon/config.py:59]: can you move this conversation to a new file call conversation.py and use this as a base class. You can inherit from it and call current one LlavaTemplate?

Comment (yashaswikarnati) [reply nemo/collections/multimodal/data/energon/config.py]: done

**Review by yaoyu-33 — APPROVED** [review]
